### PR TITLE
Fix video capture for 32/64 arch

### PR DESCRIPTION
**Pull request by oraad**
This pull request includes two fixes:

Fixed:
- Structs to support 32 and 64 arch

Added:
- Support other types of Camera resolution types (i.e. discrete, continuous, stepwise)

Changed:
- RawVideoSettings was renamed V4l2Request as it reflects the actual intention of the class (making ioctl request)

Removed:
- VideoSettings was removed because it is a hacky way of making ioctl request and V4l2Request should be used instead

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1474)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1474) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1474) before we can accept your contribution.<br/><br/>:x: oraad [sign now](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1474)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/iot?pullRequest=1474) it.</sub>

File: src/devices/Interop/Unix/Libc/Interop.videodev2.struct.cs [hunk 1]
```diff
@@ -132,17 +132,43 @@ internal enum v4l2_colorspace : uint
...
+    [FieldOffset(20)]
     public uint sizeimage;
+    [FieldOffset(24)]
     public v4l2_colorspace colorspace;
+    [FieldOffset(28)]
     public uint priv;
+    [FieldOffset(32)]
+    public uint flags;
+
+    // TODO: Implement v4l2_ycbcr_encoding enum
```
Comment (krwq) [inline src/devices/Interop/Unix/Libc/Interop.videodev2.struct.cs:157]: remove the TODOs, there is no point implementing that unless we plan to do something with it.

File: src/devices/Media/VideoDevice/Devices/UnixVideoDevice.cs [hunk 2]
```diff
@@ -168,10 +168,61 @@ public override IEnumerable<(uint Width, uint Height)> GetPixelFormatResolutions
...
-            while (V4l2Struct(VideoSettings.VIDIOC_ENUM_FRAMESIZES, ref size) != -1)
+            List<IResolution> result = new List<IResolution>();
+            while (V4l2Struct(V4l2Request.VIDIOC_ENUM_FRAMESIZES, ref size) != -1)
             {
-                result.Add((size.discrete.width, size.discrete.height));
+                IResolution resolution;
+                switch ((ResolutionType)size.type)
+                {
+                    case ResolutionType.Discrete:
+
```
Comment (krwq) [inline src/devices/Media/VideoDevice/Devices/UnixVideoDevice.cs:178]: please remove the extra enters here (only one between `};` and `break;` is fine

File: src/devices/Media/VideoDevice/Resolution.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using System;
```
Comment (krwq) [inline src/devices/Media/VideoDevice/Resolution.cs:3]: enter before using statement

File: src/devices/Media/VideoDevice/Resolution.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        /// <summary>
+        /// Stepwise
+        /// </summary>
+        Stepwise = v4l2_frmsizetypes.V4L2_FRMSIZE_TYPE_STEPWISE
+    }
+
+    /// <summary>
+    /// Resolution's interface
+    /// </summary>
+    public interface IResolution
```
Comment (krwq) [inline src/devices/Media/VideoDevice/Resolution.cs:31]: please create one type per file (exception can be interop code but not here), make sure file name should match the type name

File: src/devices/Media/VideoDevice/samples/Program.cs [hunk 5]
```diff
@@ -20,7 +20,14 @@
...
 {
-    Console.Write($"{resolution.Width}x{resolution.Height} ");
+    if (resolution is DiscreteResolution discreteRes)
+    {
+        Console.Write($"{discreteRes.Width}x{discreteRes.Height} ");
+    }
+    else if (resolution is StepwiseResolution stepRes)
+    {
+        Console.Write($"[{stepRes.MinWidth}x{stepRes.MinHeight}]->[{stepRes.MaxWidth}x{stepRes.MaxHeight}], Step [{stepRes.StepWidth},{stepRes.StepHeight}] ");
+    }
```
Comment (krwq) [inline src/devices/Media/VideoDevice/samples/Program.cs:30]: should you handle continuous resolution here?

File: src/devices/Media/VideoDevice/Resolution.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        /// <summary>
+        /// Resolution's type
+        /// </summary>
+        ResolutionType Type { get; }
```
Comment (krwq) [inline src/devices/Media/VideoDevice/Resolution.cs:36]: the type is also in the C# type itself, you can do:

`if (myIResolution is DiscreteResolution)`. Would it make sense to make this class have some common utilities? i.e. MinWidth, MaxWidth, MinHeight, MaxHeight? For DiscreteResolution that would pairwise return the same value but still would make it much easier to use

Comment (krwq) [reply src/devices/Media/VideoDevice/Resolution.cs:36]: take it with the grain of salt, I'm not super familiar with this code, I just like easy user experience if possible 😄

**Review by krwq — APPROVED** [review]
Overall looks good - couple of nits. Thank you for fixing this!

**Comment by krwq** [comment]
@Ellerbach would be good if you took the second look since you're more familiar

File: src/devices/Media/VideoDevice/Devices/UnixVideoDevice.cs [hunk 7]
```diff
@@ -488,7 +539,7 @@ private void Close()
         /// <param name="request">V4L2 request value</param>
         /// <param name="struct">The struct need to be read or set</param>
         /// <returns>The ioctl result</returns>
-        private int V4l2Struct<T>(VideoSettings request, ref T @struct)
+        private int V4l2Struct<T>(int request, ref T @struct)
```
Comment (krwq) [inline src/devices/Media/VideoDevice/Devices/UnixVideoDevice.cs:542]: Why is this changed? Wasn't enum better?

Comment (oraad) [reply src/devices/Media/VideoDevice/Devices/UnixVideoDevice.cs:542]: The enum was using pre-calculated values for the request instead of getting the correct values based on the struct size.

File: src/devices/Media/VideoDevice/Resolution.cs (hunk 6 above)
Comment (oraad) [reply src/devices/Media/VideoDevice/Resolution.cs:36]: It's possible to implement it as a single class using Min and Max, with StepHeight and StepWidth  values of 0, 1, n based on the type. Only discrete resolution will not be matching the v4l2 implementation, but I suppose it should be fine.

File: src/devices/Media/VideoDevice/samples/Program.cs (hunk 5 above)
Comment (oraad) [reply src/devices/Media/VideoDevice/samples/Program.cs:30]: In the UnixVideoDevice, I was handling the continuous as as stepwise with step 1

**Review by Ellerbach — APPROVED** [review]
Thanks for providing this PR and fixing those issues. It looks good all up for me. Few cosmetic adjustments pointed out are needed (extra enter, missing ones). And that will be good to go.

**Review by krwq — APPROVED** [review]
